File: src/vrf-utils.c
/*
 Copyright (C) 2016 Hewlett-Packard Development Company, L.P.
 All Rights Reserved.

    Licensed under the Apache License, Version 2.0 (the "License"); you may
    not use this file except in compliance with the License. You may obtain
    a copy of the License at

         http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
    License for the specific language governing permissions and limitations
    under the License.
*/

/*************************************************************************//**
 * @ingroup ops_utils
 * This module contains the DEFINES and functions that comprise the ops-utils
 * library.
 *
 * @file
 * Source file for ops-utils library.
 *
 ****************************************************************************/

#include <stdio.h>
#include <unistd.h>
#include <sched.h>
#include <string.h>
#include <errno.h>

#include <assert.h>
#include "vrf-utils.h"
#include "vswitch-idl.h"
#include "openswitch-idl.h"
#include "nl-utils.h"
/*
 * Check for presence of VRF and return VRF row.
 */

const struct ovsrec_vrf *
vrf_lookup (const struct ovsdb_idl *idl, const char *vrf_name)
{
    const struct ovsrec_vrf *vrf_row = NULL;
    OVSREC_VRF_FOR_EACH (vrf_row, idl)
      {
        if (strncmp (vrf_row->name, vrf_name, OVSDB_VRF_NAME_MAXLEN) == 0)
        return vrf_row;
      }
    return NULL;
}/*vrf_lookup*/


/*
 * Return default VRF row
 */
const struct ovsrec_vrf *
get_default_vrf (const struct ovsdb_idl *idl)
{
    /** TBD, Logic can be simplified further if we can store default VRF UUID.*/
    const struct ovsrec_vrf *vrf_row = vrf_lookup(idl, DEFAULT_VRF_NAME);

    return vrf_row;
}

/*
 * Check for presence of VRF using table_id and return VRF row.
 */
const struct ovsrec_vrf *
vrf_lookup_on_table_id (const struct ovsdb_idl *idl, const int64_t table_id)
{
    const struct ovsrec_vrf *vrf_row = NULL;
    OVSREC_VRF_FOR_EACH (vrf_row, idl)
    {
        if((vrf_row->table_id) && (*(vrf_row->table_id) == table_id))
            return vrf_row;
    }
    return NULL;
}/*vrf_lookup_on_table_id*/

/*
 * Return VRF UUId
 */
const int64_t
get_vrf_uuid_from_table_id (const struct ovsdb_idl *idl, const int64_t table_id,
                            struct uuid *uuid)
{
    const struct ovsrec_vrf *vrf_row = NULL;

    if (uuid == NULL)
       return -1;

    vrf_row = vrf_lookup_on_table_id(idl, table_id);
    if (vrf_row != NULL)
    {
        *uuid = vrf_row->header_.uuid;
        return 0;
    }

    return -1;
}

/*
 * Return VRF table_id
 */
const int64_t
get_vrf_table_id_from_uuid (const struct ovsdb_idl *idl, const struct uuid *uuid)
{
    const struct ovsrec_vrf *vrf_row = NULL;

    if (uuid == NULL)
       return -1;

    vrf_row = ovsrec_vrf_get_for_uuid(idl, uuid);
    if(vrf_row != NULL)
        return *vrf_row->table_id;

    return -1;
}

/*
 * Creates a socket by entering the corresponding namespace
 */
int create_vrf_socket (const struct ovsdb_idl *idl, int64_t table_id,
                      struct vrf_sock_params *params)
{
    const struct ovsrec_vrf *vrf_row = NULL;

    vrf_row = vrf_lookup_on_table_id(idl, table_id);
    if (vrf_row != NULL)
        return nl_create_vrf_socket(vrf_row->name, params);

    return -1;
}

/*
 * Closes a socket by entering the corresponding namespace
 */
int close_vrf_socket (const struct ovsdb_idl *idl, int64_t table_id,
                      int socket_fd)
{
    const struct ovsrec_vrf *vrf_row = NULL;

    vrf_row = vrf_lookup_on_table_id(idl, table_id);
    if (vrf_row != NULL)
    {
        if(nl_close_vrf_socket(vrf_row->name, socket_fd))
            return 0;
    }
    return -1;
}

File: vrf_utils.py
#!/usr/bin/env python
# Copyright (C) 2015-2016 Hewlett Packard Enterprise Development LP
# All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may
# not use this file except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
# License for the specific language governing permissions and limitations
# under the License.

import ovs.dirs
import ovs.daemon
import ovs.db.idl
import ovs.unixctl
import ovs.unixctl.server
import argparse
import ovs.vlog
import os

global_dstn_ns = ''
vlog = ovs.vlog.Vlog("vrf_utils")

SYSTEM_TABLE = "System"
PORT_TABLE = "Port"
VRF_TABLE = "VRF"

DEFAULT_VRF_NAME = "vrf_default"
SWITCH_NAMESPACE = "swns"
NAMESPACE_NAME_PREFIX = "VRF_"

def  get_mgmt_ip(idl):
    '''
    get the ip address of management interface
    '''
    mgmt_ip = None

    for ovs_rec in idl.tables[SYSTEM_TABLE].rows.itervalues():
        if ovs_rec.mgmt_intf_status and ovs_rec.mgmt_intf_status is not None:
            for key, value in ovs_rec.mgmt_intf_status.iteritems():
                if key == "ip":
                    mgmt_ip = value

    return mgmt_ip

def extract_ip(ip):
    '''
    extract the ip from [u'x.x.x.x/y'] format
    '''
    tmp = str(ip).split('\'')
    tmp = tmp[1].split('/')
    ip = tmp[0]

    return ip

def get_ip_from_interface_name(idl, source_interface):
    '''
    get the ip address configured on the interface
    '''
    ip = None

    for ovs_rec in idl.tables[PORT_TABLE].rows.itervalues():
        if ovs_rec.name == source_interface:
            ip = ovs_rec.ip4_address
            if len(ip) == 0:
                return None
            ip = extract_ip(ip)

    return ip

def get_vrf_name_from_ip(idl, src_ip):
    '''
    get the vrf configured on the interface
    '''
    vrf_name = None

    for ovs_rec in idl.tables[PORT_TABLE].rows.itervalues():
        ip = ovs_rec.ip4_address

        if len(ip) == 0:
            continue

        ip = extract_ip(ip)
        if ip == src_ip:
            vrf_id = ovs_rec.uuid

            for ovs_rec in idl.tables[VRF_TABLE].rows.itervalues():
                for port_row in ovs_rec.ports:

                    if port_row.uuid == vrf_id:
                        vrf_name = ovs_rec.name
                        break
    return vrf_name

def get_source_interface(idl, protocol):
    '''
    get the source interface configured for the protocol
    '''
    source_interface = None

    for ovs_rec in idl.tables["VRF"].rows.itervalues():

        if ovs_rec.name == DEFAULT_VRF_NAME:
            ip = ovs_rec.source_ip

            source_interface = ovs_rec.source_interface

            for key, value in ovs_rec.source_ip.iteritems():
                vlog.info("IP: key = %s, value = %s \n" %(str(key), str(value)))

                if key == protocol:
                    if len(value) != 0:
                        return value

                if key == "all":
                    if len(value) != 0:
                        return value

            for key, value in ovs_rec.source_interface.iteritems():

                if key == protocol:
                    return (extract_ip(value.ip4_address))

                if key == "all":
                    return (extract_ip(value.ip4_address))


def get_namespace_from_vrf(idl, vrf_name):
    '''
    get the namespace corresponding to the VRF
    get clarification from Krishna's team on this logic
    '''

    namespace = None

    for ovs_rec in idl.tables[VRF_TABLE].rows.itervalues():
         if ovs_rec.name == vrf_name:
                table_id = ovs_rec.table_id
                table_id = str(table_id).strip('[]')
                if table_id == '0':
                    namespace = SWITCH_NAMESPACE
                else:
                    namespace = NAMESPACE_NAME_PREFIX + str(table_id)

    return namespace
